cmd/keycmd/list.go

```diff
@@ -24,107 +26,351 @@ import (
 		if strings.HasSuffix(f.Name(), constants.KeySuffix) {
 			keyPaths[i] = filepath.Join(app.GetKeyDir(), f.Name())
 		}
 	}
-	return printAddresses(keyPaths)
+	addrInfos := []addressInfo{}
+	for _, keyPath := range keyPaths {
+		keyAddrInfos, err := getStoredKeyInfo(pClients, cClients, networks, keyPath, cchain)
+		if err != nil {
+			return nil, err
+		}
+		addrInfos = append(addrInfos, keyAddrInfos...)
+	}
+	return addrInfos, nil
 }
 
-func printAddresses(keyPaths []string) error {
-	header := []string{"Key Name", "Chain", "Address", "Balance", "Network"}
-	table := tablewriter.NewWriter(os.Stdout)
-	table.SetHeader(header)
-	table.SetRowLine(true)
-	table.SetAutoMergeCellsByColumnIndex([]int{0, 1})
+func getStoredKeyInfo(
+	pClients map[models.Network]platformvm.Client,
+	cClients map[models.Network]ethclient.Client,
+	networks []models.Network,
+	keyPath string,
+	cchain bool,
+) ([]addressInfo, error) {
+	addrInfos := []addressInfo{}
+	for _, network := range networks {
+		networkID, err := network.NetworkID()
+		if err != nil {
+			return nil, err
+		}
+		keyName := strings.TrimSuffix(filepath.Base(keyPath), constants.KeySuffix)
+		sk, err := key.LoadSoft(networkID, keyPath)
+		if err != nil {
+			return nil, err
+		}
+		if cchain {
+			cChainAddr := sk.C()
+			addrInfo, err := getCChainAddrInfo(cClients, network, cChainAddr, "stored", keyName)
+			if err != nil {
+				return nil, err
+			}
+			addrInfos = append(addrInfos, addrInfo)
+		}
+		pChainAddrs := sk.P()
+		for _, pChainAddr := range pChainAddrs {
+			addrInfo, err := getPChainAddrInfo(pClients, network, pChainAddr, "stored", keyName)
+			if err != nil {
+				return nil, err
+			}
+			addrInfos = append(addrInfos, addrInfo)
+		}
+	}
+	return addrInfos, nil
+}
 
-	supportedNetworks := map[string]uint32{
-		models.Fuji.String(): avago_constants.FujiID,
-		/*
-			Not enabled yet
-			models.Mainnet.String(): avago_constants.MainnetID,
-		*/
+func getLedgerAddrInfos(
+	pClients map[models.Network]platformvm.Client,
+	ledgerIndices []uint,
+	networks []models.Network,
+) ([]addressInfo, error) {
+	ledgerDevice, err := ledger.New()
+	if err != nil {
+		return nil, err
+	}
+	ux.Logger.PrintToUser("*** Please provide extended public key on the ledger device ***")
+	maxIndex := math.Max(0, ledgerIndices...)
+	toDerive := int(maxIndex + 1)
+	addresses, err := ledgerDevice.Addresses(toDerive)
+	if err != nil {
+		return nil, err
+	}
+	if len(addresses) != toDerive {
+		return nil, fmt.Errorf("derived addresses length %d differs from expected %d", len(addresses), toDerive)
 	}
+	addrInfos := []addressInfo{}
+	for _, index := range ledgerIndices {
+		addr := addresses[index]
+		ledgerAddrInfos, err := getLedgerAddrInfo(pClients, index, networks, addr)
+		if err != nil {
+			return []addressInfo{}, err
+		}
+		addrInfos = append(addrInfos, ledgerAddrInfos...)
+	}
+	return addrInfos, nil
+}
 
-	if allNetworks {
-		supportedNetworks[models.Local.String()] = 0
+func getLedgerAddrInfo(
```

Comment:
I'm not sure what the difference between this function and `getLedgerAddrInfos` is. The difference isn't clear from the name.

Reply:
renamed to getLedgerIndices info and getLedgerIndexInfo